rs/sns/swap/src/types.rs

```diff
@@ -708,17 +708,63 @@ impl DirectInvestment {
 
 impl CfInvestment {
     pub fn validate(&self) -> Result<(), String> {
-        if !is_valid_principal(&self.hotkey_principal) {
-            return Err(format!(
-                "Invalid hotkey_principal {}",
-                self.hotkey_principal
-            ));
-        }
+        self.try_get_controller()?;
+
         if self.nns_neuron_id == 0 {
             return Err("Missing nns_neuron_id".to_string());
         }
         Ok(())
     }
+
+    /// Tries to get the controller, which may be either in the `controller` or `hotkey_principal` field.
+    /// If both fields are set, requires that they refer to the same principal before returning one.
+    pub fn try_get_controller(&self) -> Result<PrincipalId, String> {
+        #[allow(deprecated)] // TODO(NNS1-3198): Remove once hotkey_principal is removed
+        match (
+            self.controller,
+            crate::swap::string_to_principal(&self.hotkey_principal),
+        ) {
+            (Some(p1), Some(p2)) if p1 == p2 => Ok(p1),
+            // If hotkey_principal refers to a different principal than controller,
+            // or if neither is set, something has gone wrong.
+            (Some(_), Some(_)) => {
+                Err("Invalid NF neuron: controller and hotkey_principal do not match".to_string())
+            }
+            // If both fields are none, something has also gone wrong.
+            (None, None) => Err(
+                "Invalid NF neuron: controller is unset and hotkey_principal is invalid"
+                    .to_string(),
+            ),
+            // If only one is set, just use that one
+            (Some(p), None) => Ok(p),
+            (None, Some(p)) => Ok(p),
+        }
+    }
+}
+
+impl CfParticipant {
+    pub fn try_get_controller(&self) -> Result<PrincipalId, String> {
+        #[allow(deprecated)] // TODO(NNS1-3198): Remove once hotkey_principal is removed
+        match (
+            self.controller,
+            crate::swap::string_to_principal(&self.hotkey_principal),
+        ) {
+            (Some(p1), Some(p2)) if p1 == p2 => Ok(p1),
+            // If hotkey_principal refers to a different principal than controller,
+            // or if neither is set, something has gone wrong.
+            (Some(_), Some(_)) => Err(
+                "Invalid NF participant: controller and hotkey_principal do not match".to_string(),
+            ),
+            // If both fields are none, something has also gone wrong.
+            (None, None) => Err(
+                "Invalid NF participant: controller and hotkey_principal are both unset"
+                    .to_string(),
+            ),
+            // If only one is set, just use that one
+            (Some(p), None) => Ok(p),
+            (None, Some(p)) => Ok(p),
+        }
+    }
```

Comment:
Note to the reviewer: this code will get much simpler once hotkey_principal is removed, so I think we should tolerate the duplication/hairiness for now. It allows this function to be tolerant to either controller being unset or `hotkey_principal` being invalid (e.g. an empty string), which will be very helpful when removing `hotkey_principal`.